### PR TITLE
Nix configuration with text-2.0

### PR DESCRIPTION
**Pull request by poorlyknitsweater**
This PR addresses #41. It
- Introduces a Nix configuration for `alfred-margaret`
  - Works in CI
  - Replaces the placeholder Nix configuration in the `benchmark` folder
- Bumps `text` to version 2.0
  - Replaces any now obsolete code in `src/Data/Text/Utf8.hs`
  - Other small changes around the codebase, e.g. `fromByteList` in `tests/Data/Text/Utf8/AhoCorasickSpec.hs`
- Ensures that code in the `benchmark` folder still runs.

### CODE REVIEW

**Comment by poorlyknitsweater** [comment]
Before merging this we should maybe tag and release the package and then bump the major version.

Instead of removing UTF-16 support wholesale we could also change the version constraint to something like `(text >= 1.2 && text < 1.3) || text >= 2.0` and then hack both versions into the same code files using `MIN/MAX_VERSION` macros.

**Comment by poorlyknitsweater** [comment]
- [ ] Update `tested-with:`
- [ ] Rename `Utf8` modules

File: benchmark/README.md
Comment (fatho) [inline benchmark/README.md:76]: Was there a problem with this approach? It looks like the intention was that it's possible to just  copy&paste this command, but I presume that breaks when there's more than one match?

**Review by fatho — COMMENTED** [review]
Nice cleanup! Just some minor things:

File: benchmark/README.md
Comment (poorlyknitsweater) [reply benchmark/README.md:32]: Yes, since the Java, Rust and Python benchmarks still expect their inputs to be UTF-16. I'd vote to create and issue to update those after the release.

File: benchmark/README.md
Comment (poorlyknitsweater) [reply benchmark/README.md:76]: Yes, the Nix config broke this (previously pasteable) line.

File: benchmark/README.md
Comment (poorlyknitsweater) [reply benchmark/README.md:86]: Neat! I will update that.

File: benchmark.sh
Comment (poorlyknitsweater) [reply benchmark.sh:1]: Did that already and fixed all the complaints :D

File: src/Data/Text/AhoCorasick/Searcher.hs
Comment (poorlyknitsweater) [reply src/Data/Text/AhoCorasick/Searcher.hs:155]: That's weird and of course not on purpose. I'll fix it.

File: alfred-margaret.cabal
Comment (poorlyknitsweater) [reply alfred-margaret.cabal:22]: Good point, I will test them all.

Comment (poorlyknitsweater) [reply alfred-margaret.cabal:20]: It does not build any more with GHC 8.6.5 from Nix so I removed that. The rest works.

File: src/Data/Text/Utf8.hs
Comment (Bodigrim) [inline src/Data/Text/Utf8.hs:293]: It is likely faster to use [`iterArray`](https://hackage.haskell.org/package/text-2.0/docs/Data-Text-Unsafe.html#v:iterArray).

File: src/Data/Text/Utf8.hs
Comment (Bodigrim) [inline src/Data/Text/Utf8.hs:149]: It's likely faster to use [`utf8Length`](https://hackage.haskell.org/package/text-2.0/docs/Data-Text-Internal-Encoding-Utf8.html#v:utf8Length) and then switch between `ord2`, `ord3`, `ord4`.

File: src/Data/Text/Utf8.hs
Comment (Bodigrim) [inline src/Data/Text/Utf8.hs:206]: It's likely faster to use [`utf8LengthByLeader`](https://hackage.haskell.org/package/text-2.0/docs/Data-Text-Internal-Encoding-Utf8.html#v:utf8LengthByLeader) and then switch between `chr2`, `chr3`, `chr4`.

**Review by Bodigrim — COMMENTED** [review]
Sorry for unsolicited review, but I'm curious if it helps to beat UTF-16 in all scenarios.

File: src/Data/Text/Utf8.hs
Comment (poorlyknitsweater) [reply src/Data/Text/Utf8.hs:293]: First of all, thank you for your review! Hope you liked the article.

We compared against `iterArray` in an earlier iteration and came to the conclusion that it is slower. Looking into its source code, this is probably because it uses [`unsafeIndex`](https://hackage.haskell.org/package/text-2.0/docs/src/Data.Text.Array.html#unsafeIndex) which performs bounds checks on the `ByteArray`. Core-2-Core optimizations will likely turn this into a single bounds check, however, `unsafeIndexCodePoint'` performs none. There was no noticeable difference in performance between using `<` and bitwise operations to check the MSBs of the first code unit.

To make sure that we aren't missing out, I just ran our benchmark against three versions of the UTF-8 code:

```
$ git log --no-decorate --oneline | head -n 3
78fc4bc Replace unsafeIndexCodePoint' by iterArray
0ab379e Remove redundant remaining parameter
22dff79 Re-add isCaseInvariant function
```

- 22dff79 is the current state of this PR
- 0ab379e removes the `remaining` parameter from the functions in the inner loop which is remnant of the UTF-16 version.
- 78fc4bc uses `iterArray` instead of our own function

The results are:

```
haskell-22dff79.stats:
  mean time: 2.726 ± 0.027 seconds
   min time: 2.618 seconds

haskell-0ab379e.stats:
  mean time: 2.505 ± 0.013 seconds
   min time: 2.417 seconds

haskell-78fc4bc.stats:
  mean time: 3.261 ± 0.011 seconds
   min time: 3.194 seconds
```

As you can see, `iterArray` is quite a bit slower. However, thanks to your review, I removed the `remaining` parameter which gave us another small improvement! I will cherry-pick that commit into this PR :rocket: 

I also took a superficial look at the generated Core, suspecting that the conversion to and from `CodeUnitIndex` could slow the `iterArray` version down. That's not the case though; it seems like the bounds checks are the problem indeed.

File: src/Data/Text/Utf8.hs
Comment (poorlyknitsweater) [reply src/Data/Text/Utf8.hs:149]: `unicode2utf8` and `decodeUtf8` are used in our test suite only, they are optimized for legibility rather than speed. This probably warrants a comment or we could move them to the `Orphans` module in `./test`.

File: src/Data/Text/Utf8.hs
Comment (poorlyknitsweater) [reply src/Data/Text/Utf8.hs:293]: I took another look at the Core and, it seems like the main difference lies in how the length of the code unit sequence constituting a code point is determined:

For `unsafeIndexCodePoint'`, the ASCII case looks like this:

```haskell
-- ipv4_sjGM is the array
-- ww2_sjHC is our current index
case indexWord8Array# ipv4_sjGM ww2_sjHC of r#1_sjHG
{ __DEFAULT ->
case ltWord# r#1_sjHG 192## of {
  __DEFAULT -> -- Check other cases
  1# -> -- Code point consists of a single code unit, use that
}
```

For `iterArray`, it looks like this:

```haskell
-- ipv4_sGDw is the array
-- ww2_sGEp is our current index
case indexWord8Array# ipv4_sGDw ww2_sGEp of r#1_sGEt
{ __DEFAULT ->
case xor# r#1_sGEt 255## of sat_sGEv { __DEFAULT ->
case clz8# sat_sGEv of sat_sGEw { __DEFAULT ->
case word2Int# sat_sGEw of c#1_sGEu { __DEFAULT ->
case <=# c#1_sGEu 0# of sat_sGEx { __DEFAULT ->
case xorI# c#1_sGEu sat_sGEx of ds_sGEy {
  __DEFAULT -> -- 4 code units
  1# -> -- 1 code unit
  2# -> -- 2 code units
  3# -> -- 3 code unit
}
```

My assumption is that since we mostly handle ASCII data, the additional operations for each character cause the overhead. For CJK or non-BMP texts `iterArray` would probably be faster since it uses less jumps. Perhaps we will try to measure that in another micro-benchmark, e.g. ASCII vs. CJK vs. non-BMP vs. random strings.

Comment (Bodigrim) [reply src/Data/Text/Utf8.hs:293]: @poorlyknitsweater Could you please try replacing all occurrences of `chr` with https://hackage.haskell.org/package/text-2.0/docs/Data-Text-Internal-Unsafe-Char.html#v:unsafeChr8 and similar? The reason is that `chr` involves a redundant safety check: https://hackage.haskell.org/package/base-4.16.1.0/docs/src/GHC-Char.html#chr

Comment (poorlyknitsweater) [reply src/Data/Text/Utf8.hs:293]: Thanks again! I tried this out, but the benchmark doesn't show a significant difference :(

I compared 952e432 (from yesterday) and 7b8be51, which replaces `chr` by either `unsafeChr8` or `unsafeChr`:

```
haskell-952e432.stats:
  mean time: 2.792 ± 0.016 seconds
   min time: 2.681 seconds

haskell-7b8be51.stats:
  mean time: 2.774 ± 0.015 seconds
   min time: 2.675 seconds
```

Both Core and CMM outputs show that the jump was actually eliminated:

952e432:

```haskell
case ltWord# r#1_s11cE 192## of {
  __DEFAULT -> -- Not ASCII
  1# -> -- ASCII
    case word2Int# r#1_s11cE of i#_s11eg { __DEFAULT -> -- These two lines are not present anymore in -ddump-opt-cmm
    case int2Word# i#_s11eg of sat_s11eh { __DEFAULT ->
    case leWord# sat_s11eh 1114111## of { -- Unnecessary check introduced by chr
      __DEFAULT -> ($wlvl i#_s11eg) `cast` <Co:4>; -- error that's never thrown
      1# ->
        case chr# i#_s11eg of ww4_s11ej { __DEFAULT -> -- Convert to ww4_s11ej :: Char#
        case ww3_s11cC of wild4_s11ek { -- Switch on current aho-corasick state
```

7b8be51:

```haskell
case ltWord# r#1_s10r6 192## of {
  __DEFAULT -> -- Not ASCII
  1# -> -- ASCII
    case word2Int# r#1_s10r6 of sat_s10sD { __DEFAULT -> -- Convert to Int# (Haskell source uses unsafeChr8 here)
    case chr# sat_s10sD of ww4_s10sC { __DEFAULT -> -- Convert to Char#
    case ww3_s10r4 of wild4_s10sE { -- Switch on current aho-corasick state
```

So it's kind of weird that this doesn't improve the benchmark times, my assumption is that this branch is just very predictable since it's never taken.

**Review by marcin-rzeznicki — APPROVED** [review]
Looked at the unreviewed part of work and glanced over the rest of it. LGTM!

File: src/Data/Text/Utf8.hs
Comment (poorlyknitsweater) [reply src/Data/Text/Utf8.hs:293]: @Bodigrim we will merge this PR now, if you have any further ideas for improvement, please create an issue! Thank you again for the review :)

**Comment by poorlyknitsweater** [comment]
@OpsBotPrime merge

**Comment by OpsBotPrime** [comment]
Pull request approved for merge by @poorlyknitsweater, rebasing now.

**Comment by OpsBotPrime** [comment]
Rebased as d1ac105380e51df7bc7c48abce5b9de09dcc939a, waiting for CI …

**Comment by Bodigrim** [comment]
@poorlyknitsweater great stuff! Thanks for checking my suggestions, I think your analysis about branch prediction is right.

**Pull request merged**
